src/write.rs

```diff
@@ -42,53 +47,100 @@ impl<W: Write> BrotliEncoder<W> {
         let mut data = Compress::new();
         data.set_params(params);
         BrotliEncoder {
-            max: data.input_block_size(),
-            cur: 0,
             data: data,
-            obj: Some(obj)
+            obj: Some(obj),
+            buf: Vec::with_capacity(BUF_SIZE),
+            cur: 0,
+            err: None,
         }
     }
 
 
-    fn do_finish(&mut self) -> io::Result<()> {
-        let data = try!(self.data.compress(true, false));
-        self.obj.as_mut().unwrap().write_all(data)
+    fn dump(&mut self) -> io::Result<()> {
+        loop {
+            while !self.buf.is_empty() {
+                let amt = try!(self.obj.as_mut().unwrap().write(&self.buf[self.cur..]));
+                self.cur += amt;
+                if self.cur == self.buf.len() {
+                    self.buf.clear();
+                    self.cur = 0
+                }
+            }
+            // TODO: if we could peek, the buffer wouldn't be necessary
```

Comment:
We could do an opportunistic write here, right? And then only buffered unwritten data?